api/app/services/privacy_retention_service.py

```
@@ -0,0 +1,1930 @@
+        referenced_ids: set[int] = set()
+        if old_ids:
+            placeholders = ",".join("?" for _ in old_ids)
+            for reference_table in (
+                "conversation_threads",
+                "knowledge_update_proposals",
+            ):
+                if not self._table_exists(connection, reference_table):
+                    continue
+                reference_rows = connection.execute(
+                    f"SELECT DISTINCT candidate_id FROM {reference_table} "
+                    f"WHERE candidate_id IN ({placeholders})",
```

Ruff (0.15.21)

[error] Possible SQL injection vector through string-based query construction

(S608)